### ISSUE TITLE
Behat test suite is porous

### ISSUE THREAD

**Issue by unclecheese** [silverstripe/silverstripe-framework#8112]
The behat tests we run on framework are all dependent on the `silverstripe-admin` module. It's possible to get changes merged into `admin`, get a clean CI, merge, and then unknowingly break the framework tests, which were never run.

The only foolproof way around this is to have `silverstripe-admin` run behat tests on all core modules, but that seems expensive and difficult to maintain. Plus, the last thing we need is a slower CI workflow.

Open for discussion...

## Pull requests

- [x] https://github.com/silverstripe/silverstripe-framework/pull/8172
- [x] https://github.com/silverstripe/silverstripe-admin/pull/540
- [x] https://github.com/silverstripe/silverstripe-admin/pull/547

**Comment by robbieaverill** [comment]
I see the benefit in moving the tests, but I disapprove of renaming them. Any module that uses behat will probably use these contexts and will now break on this core version, which will require a new release in order to change the fixture names. I realise it’s not mission critical but it’s still a breaking change and is a pain. Perhaps we could move the tests but leave the namespaces as they were?

**Comment by unclecheese** [comment]
Fair enough. I guess I don't consider the tests part of the API, but I take your point that it could cause pain for consumers of the core Behat contexts.

I do think we need to be mindful of decoupling and keeping things tidy. As an alternative, rather than deal with the dissonance of a `\Framework\` class in `Admin`, why not just move these classes back into `framework`, which will not run any behat tests itself, but be the provider of core behat features for modules that do want to consume them.

**Comment by robbieaverill** [comment]
> I guess I don't consider the tests part of the API, but I take your point that it could cause pain for consumers of the core Behat contexts.

I don't really either, but yeah it will be a pain for module maintainers (like us)

I like your suggestion for keeping the context classes in framework but not running any tests on it!

**Comment by maxime-rainville** [comment]
I'll hold off peer reviewing this until @unclecheese has done is magic.

**Comment by unclecheese** [comment]
Closing per discussion in https://github.com/silverstripe/silverstripe-framework/issues/8112#issuecomment-397064004

**Comment by maxime-rainville** [comment]
@tractorcow helpfully pointed out that most of our modules running behat test reference travis-upload-artifacts.php 5 seconds after I merged the PR.

They will call that file if the behat test fails.
```bash
if [[ $BEHAT_TEST ]]; then php vendor/silverstripe/framework/tests/behat/travis-upload-artifacts.php --if-env BEHAT_TEST,ARTIFACTS_BUCKET,ARTIFACTS_KEY,ARTIFACTS_SECRET --target-path $TRAVIS_REPO_SLUG/$TRAVIS_BUILD_ID/$TRAVIS_JOB_ID --artifacts-base-url https://s3.amazonaws.com/$ARTIFACTS_BUCKET/ --artifacts-path ./artifacts/; fi
```

**Comment by tractorcow** [comment]
@unclecheese don't forget to rename the fixture class names in the behat.yml files.

https://github.com/silverstripe/silverstripe-siteconfig/blob/4/behat.yml

**Comment by maxime-rainville** [comment]
I've created https://github.com/silverstripe/silverstripe-admin/pull/547 to revert the behavior for uploading behat test result to S3 on failure.

I've restored `travis-upload-artifacts.php` on framewor and accidentally pushed it directly to 4 without a PR. https://github.com/silverstripe/silverstripe-framework/commit/c3b556d2464346e753842a9e57d71db0eb7b03ae

**Comment by maxime-rainville** [comment]
This one is a gift that just keeps on giving.

Looks like we broke behat test https://travis-ci.org/silverstripe/silverstripe-installer/builds/394381909?utm_source=slack&utm_medium=notification

We'll need to restore behat.yml

**Comment by unclecheese** [comment]
Have added it back in https://github.com/silverstripe/silverstripe-framework/pull/8201, with pathing to admin/ for files and features, so it's still technically not a behat provider.

**Comment by maxime-rainville** [comment]
Cool. I've manually restarted the silverstripe-installer build. Fingers crossed.

https://travis-ci.org/silverstripe/silverstripe-installer/builds/394381909?utm_source=slack&utm_medium=notification.

**Comment by dhensby** [comment]
behat tests are red

**Comment by tractorcow** [comment]
You could also change the @framework to @admin

**Comment by chillu** [comment]
@maxime-rainville Can you check if this can be closed please?
